File: packages/javascript/lib/string/capitalizeEachWords.ts
import { contains } from '../verify/contains.js';
import { capitalizeFirst } from './capitalizeFirst.js';

export function capitalizeEachWords(str: string, natural?: boolean): string {
	if (!str) {
		return '';
	}

	let tempStr: string = str.trim();

	if (natural) {
		tempStr = tempStr.toLowerCase();
	}

	const splitStr = tempStr.split(' ');

	for (let i = 0, iLen = splitStr.length; i < iLen; i += 1) {
		if (
			!natural ||
			!contains(
				splitStr[i],
				[
					'in',
					'on',
					'the',
					'at',
					'and',
					'or',
					'of',
					'for',
					'to',
					'that',
					'a',
					'by',
					'it',
					'is',
					'as',
					'are',
					'were',
					'was',
					'nor',
					'an'
				],
				true
			)
		) {
			splitStr[i] = capitalizeFirst(splitStr[i]);
		}
	}

	return capitalizeFirst(splitStr.join(' '));
}

File: packages/python/qsu/string/capitalizeEachWords.py
from .capitalizeFirst import capitalizeFirst

# A set, so the lookup below is O(1) instead of a linear scan through `contains`.
_STOP_WORDS = frozenset([
	'in',
	'on',
	'the',
	'at',
	'and',
	'or',
	'of',
	'for',
	'to',
	'that',
	'a',
	'by',
	'it',
	'is',
	'as',
	'are',
	'were',
	'was',
	'nor',
	'an',
])


def capitalizeEachWords(str: str, natural: bool = False) -> str:
	if not str:
		return ''

	tempStr = str.strip()

	if natural:
		tempStr = tempStr.lower()

	splitStr = tempStr.split(' ')

	for i in range(len(splitStr)):
		if not natural or splitStr[i] not in _STOP_WORDS:
			splitStr[i] = capitalizeFirst(splitStr[i])

	return capitalizeFirst(' '.join(splitStr))
